### ISSUE TITLE
Broken test runner when narrowing tests

### ISSUE THREAD

**Issue by tinganho** [Microsoft/TypeScript#3273]
When running `jake runtests` with the option `t=` there is maximum call stack error.

```typescript
$ jake runtests t=amdImportAsPrim
rm -rf tests/baselines/local/
rm -rf internal/baselines/rwc/local
Running test(s): amdImportAsPrim
mocha -R dot -g amdImportAsPrim --colors  -t 20000 built/local/run.js


  ․․․․․․

  5 passing (159ms)
  1 failing

  1) compiler tests for tests/cases/conformance/externalModules/amdImportAsPrimaryExpression.ts "after all" hook:
     Uncaught RangeError: Maximum call stack size exceeded
```
The above is from the current master with an arbitrary test file.
```
node -v
v0.12.2
```

**Comment by DanielRosenwasser** [comment]
I think @RyanCavanaugh might have mentioned something about this in passing.

**Comment by danquirk** [comment]
@ahejlsberg was actually the one running into this. We debugged through it and found the harness was actually in a different test than what it was reporting when failing (if you look at the after hook in the harness it's just not doing anything that could stack overflow). It ended up being in a suspicious sounding test given Anders changes' and I didn't hear about the final resolution which I assumed meant things were back to normal (ie working). Apparently we need to investigate further.

**Comment by tinganho** [comment]
This bug was fixed and released in Mocha recently.

**Comment by mhegazy** [comment]
thanks @tinganho

**Comment by danquirk** [comment]
@tinganho out of curiousity, what was the mocha issue?

**Comment by tinganho** [comment]
@danquirk there was a recursive loop that traversed all the tests in Mocha. But when we passed a grep option that recursive loop was recursing without a next tick on the event loop, which caused the maximum call stack error. 

Here was the commit:
https://github.com/mochajs/mocha/commit/1a68cc2f102dcf73dc53515aed9112360632b1b4
